### PR TITLE
Freemabd1/rw 114 Added persist functionality for enums when creating CohortAnnotationDefinitions.

### PR DESCRIPTION
**Pull request by freemabd1**
Added persist functionality for enums when creating CohortAnnotationDefinitions.

### CODE REVIEW

File: api/db/changelog/db.changelog-23.xml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                                 constraintName="fk_enum_cohort_annotation_definition_id"
+                                 referencedColumnNames="cohort_annotation_definition_id"
+                                 referencedTableName="cohort_annotation_definition"/>
+
+        <dropColumn columnName="annotation_value_timestamp" tableName="participant_cohort_annotations"/>
+        <addColumn tableName="participant_cohort_annotations">
+            <column name="annotation_value_date" type="date"/>
+        </addColumn>
+        <addColumn tableName="participant_cohort_annotations">
+            <column name="annotation_value_enum" type="varchar(400)"/>
```
Comment (danqrodney) [inline api/db/changelog/db.changelog-23.xml]: Shouldn't this be a FK to cohort_annotation_enum_value?

File: api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java [hunk 2]
```diff
@@ -78,6 +83,30 @@ public CohortAnnotationDefinition annotationType(AnnotationType annotationType)
...
+    @OneToMany(fetch = FetchType.EAGER, mappedBy = "cohortAnnotationDefinition", orphanRemoval = true, cascade = CascadeType.ALL)
+    public Set<CohortAnnotationEnumValue> getEnumValues() {
+        return enumValues;
+    }
+
+    public void setEnumValues(Set<CohortAnnotationEnumValue> enumValues) {
+        this.enumValues = enumValues;
+        if (enumValues != null) {
+            for (CohortAnnotationEnumValue enumValue : enumValues) {
+                enumValue.setCohortAnnotationDefinition(this);
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java]: So I think the right way of dealing with collections in JPA is to do:

getEnumValues().add(foo) or remove(foo)

like we do with WorkspaceUserRole here:

https://github.com/all-of-us/workbench/blob/5c20fc2700ef7df3f089ac2bfd2de48adbef3775/api/src/main/java/org/pmiops/workbench/db/dao/WorkspaceServiceImpl.java#L160

The setter for the collection doesn't actually work properly with JPA.

That avoids this weirdness with setting this on the children, anyway...

File: api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java [hunk 3]
```diff
@@ -78,6 +83,30 @@ public CohortAnnotationDefinition annotationType(AnnotationType annotationType)
...
+            }
+        }
+    }
+
+    public CohortAnnotationDefinition enumValues(Set<CohortAnnotationEnumValue> enumValues) {
+        this.enumValues = enumValues;
+        if (enumValues != null) {
+            for (CohortAnnotationEnumValue enumValue : this.enumValues) {
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java]: Same as above? Shouldn't this just call setEnumValues()?

File: api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java [hunk 4]
```diff
@@ -78,6 +83,30 @@ public CohortAnnotationDefinition annotationType(AnnotationType annotationType)
         return this;
     }
 
+    @OneToMany(fetch = FetchType.EAGER, mappedBy = "cohortAnnotationDefinition", orphanRemoval = true, cascade = CascadeType.ALL)
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java]: You may want to make this a SortedSet with OrderBy like here:

https://stackoverflow.com/questions/19871765/jpa-hibernate-sorted-collection-orderby-vs-sort

File: api/src/main/java/org/pmiops/workbench/api/CohortAnnotationDefinitionController.java [hunk 5]
```diff
@@ -35,11 +40,14 @@
             new Function<org.pmiops.workbench.db.model.CohortAnnotationDefinition, CohortAnnotationDefinition>() {
                 @Override
                 public CohortAnnotationDefinition apply(org.pmiops.workbench.db.model.CohortAnnotationDefinition cohortAnnotationDefinition) {
+                    List<String> enumValues = cohortAnnotationDefinition.getEnumValues() == null ? null :
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/CohortAnnotationDefinitionController.java]: Are these intended to be ordered by the order field?

File: api/src/main/java/org/pmiops/workbench/api/CohortAnnotationDefinitionController.java [hunk 6]
```diff
@@ -48,10 +56,17 @@ public CohortAnnotationDefinition apply(org.pmiops.workbench.db.model.CohortAnno
             new Function<CohortAnnotationDefinition, org.pmiops.workbench.db.model.CohortAnnotationDefinition>() {
                 @Override
                 public org.pmiops.workbench.db.model.CohortAnnotationDefinition apply(CohortAnnotationDefinition cohortAnnotationDefinition) {
+                    List<String> enumValues = cohortAnnotationDefinition.getEnumValues();
+                    Set<CohortAnnotationEnumValue> enumValuesSet = (enumValues == null) ? null :
+                            new HashSet<CohortAnnotationEnumValue>(
+                                    IntStream.range(0, enumValues.size())
+                                            .mapToObj(i -> new CohortAnnotationEnumValue().name(enumValues.get(i)).order(i))
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/CohortAnnotationDefinitionController.java]: How do you update an existing enum or enum value? Are these immutable once they are created?

File: api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java (hunk 4 above)
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java]: Changed it to order by primary key.

File: api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java (hunk 2 above)
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java]: I can change it this if you like or i could just add comments about why the setter is doing this...  but I do believe that another developer will come along(not understanding the relation) and call the set method and will spend time debugging why this collection isn't being persisted. In my experience this happens a lot. If i implement and add() method... It's not always clear that calling add() is the correct thing to do.

File: api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java (hunk 3 above)
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/db/model/CohortAnnotationDefinition.java]: Done.

**Pull request merged**